### PR TITLE
Mining changes

### PR DESCRIPTION
**Pull request by RugnirViking**
----------------------
**Mining Changes**

## Summary
Two changes to mining to make it viable later into a playthrough
 - added a new keybind, defaulting to "v", that causes your escorts to follow you and pick up nearby resources.
 - made it so that when you have both an asteroid scanner and a tactical scanner you can see the HP of asteroids you have targeted.

## UI Screenshots
uipanel
asteroidhp
Collect

## Testing Done
Not a great deal to be honest. The asteroid change is pretty simple, but I'd love some support testing the "collect" order. I have of course tested it in play myself and the escorts do indeed collect boxes out to a certain distance from the flagship. They do also stick with the order when jumping between systems. 

One thing I did notice is at the moment full escorts still attempt to pick up resources, but ultimately it seems harmless.

## Performance Impact
N/A

The only issues I forsee is the name on the settings panel is a little long. I chose that name because I wanted to avoid the word "gather" given that there is already an order named that (gather around flagship)

### CODE REVIEW

**Comment by a-random-lemurian** [comment]
> The only issues I forsee is the name on the settings panel is a little long. I chose that name because I wanted to avoid the word "gather" given that there is already an order named that (gather around flagship)

Maybe rename it to "Collect resources"?

**Comment by RugnirViking** [comment]
> > The only issues I forsee is the name on the settings panel is a little long. I chose that name because I wanted to avoid the word "gather" given that there is already an order named that (gather around flagship)
> 
> Maybe rename it to "Collect resources"?

I just tested it, even "Toggle collect resources" is too long with ctrl or shift as the bindings. Perhaps remove "toggle" entirely as well
lctrl

**Comment by samrocketman** [comment]
#6669 related and uses same shortcut for the same purpose

**Comment by EjoThims** [comment]
Probably best to break off the asteroid hp as a separate thing.

Not only because there's other implementations of the other feature here, but also because that is a much simpler change and they are wholly separate under the hood as well as not relying on each other in gameplay anyway.

File: source/Minable.cpp [hunk 1]
```diff
@@ -183,6 +185,12 @@ void Minable::TakeDamage(const Projectile &projectile)
 }
 
 
+double Minable::Hull() const
+{
+	return min(1., hull / maxhull);
+}
+
+
```
Comment (samrocketman) [inline source/Minable.cpp:193]: Add a third line here.

File: source/AI.cpp [hunk 2]
```diff
@@ -1348,6 +1354,17 @@ bool AI::FollowOrders(Ship &ship, Command &command) const
 		else
 			command.SetTurn(TurnToward(ship, TargetAim(ship)));
 	}
+	else if(type == Orders::COLLECT){
```
Comment (samrocketman) [inline source/AI.cpp:1357]:
```suggestion
	else if(type == Orders::COLLECT)
	{
```

File: source/AI.cpp [hunk 3]
```diff
@@ -1348,6 +1354,17 @@ bool AI::FollowOrders(Ship &ship, Command &command) const
...
+		if(DoHarvesting(ship, command))
+		{
+			ship.SetCommands(command);
+			ship.SetCommands(firingCommands);
+			return true;
+		} else {
```
Comment (samrocketman) [inline source/AI.cpp:1363]:
```suggestion
		}
		else
```

Single statements do not include curly braces in this project style.  Example,
```cpp
	else
		CircleAround(ship, command, *target);
```

File: source/AI.cpp [hunk 4]
```diff
@@ -3629,7 +3645,7 @@ void AI::MovePlayer(Ship &ship, const PlayerInfo &player, Command &activeCommand
 		{
 			PrepareForHyperspace(ship, command);
 			command |= Command::JUMP;
-			
+
```
Comment (samrocketman) [inline source/AI.cpp:3648]: Removing this is fine; though generally you only want to touch lines of code where you're changing logic.

File: source/Minable.cpp [hunk 5]
```diff
@@ -44,8 +44,10 @@ void Minable::Load(const DataNode &node)
 		// the frame rate will be set randomly and it will always be looping.
 		if(child.Token(0) == "sprite" && child.Size() >= 2)
 			SetSprite(SpriteSet::Get(child.Token(1)));
-		else if(child.Token(0) == "hull" && child.Size() >= 2)
+		else if(child.Token(0) == "hull" && child.Size() >= 2){
```
Comment (samrocketman) [inline source/Minable.cpp:47]:
```suggestion
		else if(child.Token(0) == "hull" && child.Size() >= 2)
		{
```

File: source/AI.cpp [hunk 6]
```diff
@@ -1945,7 +1963,6 @@ void AI::CircleAround(Ship &ship, Command &command, const Body &target)
 }
```
Comment (Hurleveur) [inline source/AI.cpp:1965]:
```suggestion


```

File: source/AI.cpp (hunk 4 above)
Comment (Hurleveur) [reply source/AI.cpp:3649]: I think this was added by a miss-merge, all code files should be free of useless whitespaces. 
This PR is so minuscule that I don't mind it this time, but samenrocketman is right.

File: .gitignore [hunk 7]
```diff
@@ -66,3 +66,5 @@ ipch/
 CMakeLists.txt
 compile_commands.json
 /.gitmodules
+cscope.out
+EndlessSky.cscope_file_list
```
Comment (Hurleveur) [inline .gitignore:70]:
```suggestion
```
I don't think you should be changing this here, its completely unrelated to the PR, if you want these changes for personal use you can just open a new PR for it, which will get merged super fast probably, if people agree.

File: source/AI.cpp [hunk 8]
```diff
@@ -1348,6 +1354,18 @@ bool AI::FollowOrders(Ship &ship, Command &command) const
 		else
 			command.SetTurn(TurnToward(ship, TargetAim(ship)));
 	}
+	else if(type == Orders::COLLECT)
+	{
+		if(DoHarvesting(ship, command))
```
Comment (Hurleveur) [inline source/AI.cpp:1359]: I'll be opening a PR to change this to account for staying around the flagship when there is not anymore cargo space.

**Review by Hurleveur — CHANGES_REQUESTED** [review]
A few small things to change, I'll be opening a PR to fix the fact they keep gathering even when they can't fit asteroids anymore in their cargo. (which I'd like some feedback about btw, should it be the case or not, my opinion being it ships without cargo shouldn't waste precious minerals)

**Comment by Hecter94** [comment]
[RugnirViking](https://github.com/RugnirViking) It looks like this PR still has a pending review comment, and merge conflicts as well. 
Were you still interested in working on this? I'd like to have someone take a look at it if you can resolve the merge conflicts and any outstanding review comments.

File: source/AI.cpp [hunk 9]
```diff
@@ -2463,7 +2481,7 @@ void AI::DoMining(Ship &ship, Command &command)
 
 
 
-bool AI::DoHarvesting(Ship &ship, Command &command)
+bool AI::DoHarvesting(Ship &ship, Command &command) const
```
Comment (Hurleveur) [inline source/AI.cpp:2484]: Just wondering, why was this made const? That change should probably be removed and then I'm good with this PR

**Comment by Amazinite** [comment]
Marking this as waiting on review since it's related to #6669, which is similarly waiting on reviewer.

File: source/AI.cpp [hunk 10]
```diff
@@ -1348,6 +1354,18 @@ bool AI::FollowOrders(Ship &ship, Command &command) const
...
+	{
+		if(ship.Cargo().Free() && DoHarvesting(ship, command))
+		{
+			ship.SetCommands(command);
+			ship.SetCommands(firingCommands);
+			return true;
+		}
+		else
+			// there are no harvestables to collect
+			CircleAround(ship, command, *target);
```
Comment (warp-core) [inline source/AI.cpp:1367]: What if `target` is null?

**Comment by samrocketman** [comment]
@RugnirViking do you mind if I break asteroid health UI out to a separate pull request?  I can preserve your authorship so you will still get credit for merging.

**Comment by RugnirViking** [comment]
> @RugnirViking do you mind if I break asteroid health UI out to a separate pull request? I can preserve your authorship so you will still get credit for merging.

go for it

**Comment by samrocketman** [comment]
> go for it, sorry I havent been so active

No worries!  Thanks for this contribution and allowing me to repurpose some of it.

**Comment by samrocketman** [comment]
I'll post here the new PR when it opens

**Comment by samrocketman** [comment]
I have split out the asteroid health feature into a separate PR.

**Comment by Amazinite** [comment]
Closing since this is covered by #6669 and #8558.

**Pull request closed without merging**